packages/gui/src/components/nfts/provider/hooks/useMetadataData.ts

```diff
@@ -0,0 +1,186 @@
+      const nftId = getNFTId(id);
+
+      const metadataOnDemand = metadatasOnDemand.get(nftId);
+      if (metadataOnDemand) {
+        if (metadataOnDemand.error) {
+          throw metadataOnDemand.error;
+        }
+
+        if (metadataOnDemand.metadata) {
+          return metadataOnDemand.metadata;
+        }
+
+        if (metadataOnDemand.promise) {
+          return metadataOnDemand.promise;
+        }
+      }
+
+      async function limitedFetchMetadata() {
+        try {
+          log(`Fetching metadata for ${id} from API`);
+          const nft = await fetchNFT(nftId);
+          const { metadataUris = [], metadataHash } = nft;
+
+          const [firstUri] = metadataUris;
+          if (!firstUri) {
+            throw new Error('No metadata URI');
+          }
+
+          const metadata = await fetchAndProcessMetadata(firstUri, metadataHash);
+          setMetadataOnDemand(nftId, { metadata });
+          return metadata;
+        } catch (e) {
+          setMetadataOnDemand(nftId, { error: e as Error });
+          throw e;
+        }
+      }
+
+      const promise = limitedFetchMetadata();
+
+      setMetadataOnDemand(nftId, { promise });
+
+      return promise;
+    },
+    [
+      fetchAndProcessMetadata /* immutable */,
+      fetchNFT /* immutable */,
+      metadatasOnDemand /* immutable */,
+      setMetadataOnDemand /* immutable */,
+    ]
+  );
+
+  // immutable function
+  const getMetadata = useCallback(
+    (id: string | undefined): MetadataState => {
+      if (!id) {
+        return {
+          metadata: undefined,
+          isLoading: false,
+          error: new Error('Invalid NFT id'),
+        };
+      }
+
+      const nftId = getNFTId(id);
+
+      const metadataOnDemand = metadatasOnDemand.get(nftId);
+      if (metadataOnDemand) {
+        return {
+          metadata: metadataOnDemand.metadata,
+          isLoading: !!metadataOnDemand.promise,
+          error: metadataOnDemand.error,
+        };
+      }
+
+      fetchMetadata(nftId).catch((e) => {
+        log(`Error fetching Metadata for nftId: ${nftId}`, e);
+      });
+
+      return {
+        metadata: undefined,
+        isLoading: true,
+        error: undefined,
+      };
+    },
+    [fetchMetadata /* immutable */, metadatasOnDemand /* immutable */]
+  );
+
+  // immutable function
+  const invalidate = useCallback(
+    async (id: string) => {
+      const nftId = getNFTId(id);
+
+      const metadataOnDemand = metadatasOnDemand.get(nftId);
+      if (metadataOnDemand) {
+        // wait for the promise to resolve and ignore error
+        if (metadataOnDemand.promise) {
+          await metadataOnDemand.promise.catch(() => {});
+        }
+
+        metadatasOnDemand.delete(nftId);
+
+        // reload metadata
+        getMetadata(id);
+      }
+    },
+    [getMetadata /* immutable */, metadatasOnDemand /* immutable */]
+  );
+
+  // immutable function
+  const subscribeToMetadataChanges = useCallback(
```

Comment:
Seems like this can only be (properly) used within a `useEffect` since it returns a cleanup function. Looks like the only caller is in useNFTMetadata currently, which calls it from a useEffect hook. I'd be a bit concerned about using it elsewhere inappropriately.

Reply:
useMetadataData is an internal hook used only in Provider. 
therefore it is not inside src/hooks directory